utils/watermark_disk.go

```diff
@@ -0,0 +1,105 @@
+package utils
+
+import (
+	"crypto/rand"
+	"io"
+	"log"
+	"math/big"
+	"os"
+	"slices"
+)
+
+const (
+	bufferSize    = 512
+	numWatermarks = 10
+)
+
+type watermark struct {
+	position uint64
+	data     []byte
+}
+
+func PrepWatermarks(disk string) (func() bool, error) {
+	log.Printf("%s | Initiating watermarking process", disk)
+	watermarks := make([]watermark, numWatermarks)
+
+	// Write open
+	file, err := os.OpenFile(disk, os.O_WRONLY, 0)
+	if err != nil {
+		log.Printf("%s | Failed to open disk, err: %s", disk, err)
+		return nil, err
+	}
+	defer file.Close()
+
+	// Get disk or partition size
+	fileSize, err := file.Seek(0, io.SeekEnd)
+	if err != nil {
+		log.Printf("%s | Failed to get size, err: %s", disk, err)
+		return nil, err
+	}
+	// Write watermarks on random locations
+	err = writeWatermarks(file, fileSize, watermarks)
+	if err != nil {
+		log.Printf("%s | Failed to write watermarks, err: %s", disk, err)
+		return nil, err
+	}
+
+	checker := func() bool {
+		log.Printf("%s | Checking if the watermark has been removed", disk)
+
+		file, err := os.OpenFile(disk, os.O_RDONLY, 0)
+		if err != nil {
+			log.Printf("%s | Failed to open disk for reading: %s", disk, err)
+			return false
+		}
+		defer file.Close()
+
+		for _, watermark := range watermarks {
+			_, err = file.Seek(int64(watermark.position), io.SeekStart)
+			if err != nil {
+				log.Printf("%s | Error moving rw pointer: %s", disk, err)
+				return false
+			}
+			// Read the watermark written to the position
+			currentValue := make([]byte, bufferSize)
+			_, err = file.Read(currentValue)
+			if err != nil {
+				log.Printf("%s | Error reading from file: %s", disk, err)
+				return false
+			}
+			// Check if the watermark is still in the disk
+			if slices.Equal(currentValue, watermark.data) {
+				log.Printf("%s | Error existing watermark in the file", disk)
+				return false
+			}
+		}
+		log.Printf("%s | Watermarks has been removed", disk)
+		return true
+	}
+	return checker, nil
+}
+
+func writeWatermarks(file *os.File, fileSize int64, watermarks []watermark) error {
+	for i := 0; i < numWatermarks; i++ {
+		data := make([]byte, bufferSize)
+		_, err := rand.Read(data)
+		if err != nil {
+			log.Println("error:", err)
+			return err
+		}
+		randomPosition, err := rand.Int(rand.Reader, big.NewInt(fileSize))
+		if err != nil {
+			log.Println(err)
+			return err
+		}
+
+		_, err = file.Seek(int64(randomPosition.Uint64()), io.SeekStart)
+		if err != nil {
+			log.Println("Error moving rw pointer:", err)
+			return err
+		}
+		watermarks[i].position = randomPosition.Uint64()
+		watermarks[i].data = data
+	}
+	return nil
+}
```

Comment:
```suggestion
func writeWatermarks(r *mathrand.Rand, dev io.WriteSeeker, count int, bufferSize int64) ([]watermarks, error) {
	devSize, err := dev.Seek(0, io.SeekEnd)
	if err != nil {
		return nil, err
	}

	partitionSize := int64(devSize / count)
	marks := make([]watermarks, 0, count)

	// Divide the device into count "partitions"
	// Then we write random data somewhere randomly within each partition
	// This way we spread out the writes through out the disk and avoid any groupings (random is random after all)
	for whence := int64(0); whence < devSize; whence += partitionSize {
		data := make([]byte, bufferSize)
		_, err := rand.Read(data)
		if err != nil {
			return nil, err
		}

		offset, err := r.Int64N(partitionSize - bufferSize)
		if err != nil {
			return nil, err
		}

		pos := whence + offset
		_, err = dev.Seek(pos, io.SeekStart)
		if err != nil {
			return nil, err
		}

		err = dev.Write(data)
		if err != nil {
			return nil, err
		}
		watermarks = append(watermarks, watermark{position: pos, data: data})
	}
	return nil
}
```

I've reworked this function with a couple of enhancements.

This new signature should be more amenable for testing, which we should definitely add for this functionality :D.

The other thing it does is makes sure to divide the disk into count chunks/partitions so that we are sure to spread the watermarks all throughout the disks. Since the previous code was just getting random positions all over the disks there could be a chance that we were writing to only a small section of the disk, or even having overlapping writes. I also tweaked the random position calculation to make sure we have space for the full write w/o crossing partitions or the end-of-disk.

Reply:
Also not a big deal but maybe worth discussing. Instead of saving the data we can save a sha256 or w/e hash of the data instead. This way we can easily grow the data size without having to worry about memory usage.

Reply:
I was thinking more bytes not more marks and we will be running this in parallel against each disk but yeah not a big deal. 

https://pkg.go.dev/math/rand/v2#Rand.Int64N